File: 4/4-2.py
with open("input.txt", "r") as file:
    data = file.read()

with open("example.txt", "r") as file:
    t_data = file.read()


directions = [(-1,1), (1,1)]
total_x_mas = 0

matrix = [list(row) for row in data.strip().split("\n")]
rows, cols = len(matrix), len(matrix[0])

for i in range(rows):
    for j in range(cols):
        
        if matrix[i][j] == "A":

            valid = True

            for dir in directions:
                dir_x = i + dir[0]
                dir_y = j + dir[1]
                m_dir_x = i - dir[0]
                m_dir_y = j - dir[1]

                if dir_x < 0 or dir_x >= rows or dir_y < 0 or dir_y >= cols \
                or m_dir_x < 0 or m_dir_x >= rows or m_dir_y < 0 or m_dir_y >= cols:
                    valid = False
                    break

                p_diag = matrix[dir_x][dir_y]
                n_diag = matrix[m_dir_x][m_dir_y]

                if not ((p_diag == "M" and n_diag == "S") or (p_diag == "S" and n_diag == "M")):
                    valid = False
                    break
            
            if valid: total_x_mas += 1

print(total_x_mas)

File: 2025/4/4-2.ts
import * as fs from "fs";

const data = fs.readFileSync("input.txt", "utf8");
const lines = data.split(/\r?\n/).map((l) => l.split(""));

const DIRECTIONS = [
    [0,1], [1,0], [1,1], [0, -1], 
    [-1, 0], [-1, -1], [1, -1], [-1, 1]
];
const rows = lines.length;
const cols = lines[0].length;

const inBounds = (row: number, col: number) => row >= 0 && row < rows && col >= 0 && col < cols;

const check = (r: number, c: number, matrix: String[][]) => {
    if (matrix[r][c] !== "@") return false;
    let count = 0
    for (const [dr, dc] of DIRECTIONS) {
        const nr = r + dr, nc = c + dc;
        if (!inBounds(nr, nc)) continue;
        if (matrix[nr][nc] === '@') count ++;
        if (count > 3) return false;
    }
    return true;
}

let res=0;
while (true) {
    let found = false;

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            if (check(r, c, lines)) {
                lines[r][c] = "."
                res++;
                found = true;
            };
        }   
    }
    if (!found) break;
}

console.log(res);